cmd/ceremony/main.go

```diff
@@ -500,48 +620,151 @@ func rootCeremony(configBytes []byte) error {
+		return err
 	}
 	issuer, err := loadCert(config.Inputs.IssuerCertificatePath)
 	if err != nil {
 		return fmt.Errorf("failed to load issuer certificate %q: %s", config.Inputs.IssuerCertificatePath, err)
 	}
-
+	toBeCrossSigned, err := loadCert(config.Inputs.CertificateToCrossSignPath)
+	if err != nil {
+		return fmt.Errorf("failed to load toBeCrossSigned certificate %q: %s", config.Inputs.CertificateToCrossSignPath, err)
+	}
 	signer, randReader, err := openSigner(config.PKCS11, issuer.PublicKey)
 	if err != nil {
 		return err
 	}
-
-	template, err := makeTemplate(randReader, &config.CertProfile, pubPEM.Bytes, ct)
+	template, err := makeTemplate(randReader, &config.CertProfile, pubBytes, toBeCrossSigned, ct)
 	if err != nil {
 		return fmt.Errorf("failed to create certificate profile: %s", err)
 	}
 	template.AuthorityKeyId = issuer.SubjectKeyId
-
-	err = signAndWriteCert(template, issuer, pub, signer, config.Outputs.CertificatePath, config.SkipLints)
+	lintCert, err := issueLintCertAndPerformLinting(template, issuer, pub, signer, config.SkipLints)
 	if err != nil {
 		return err
 	}
+	// Ensure that we've configured the correct certificate to cross-sign compared to the profile.
+	//
+	// Example of a misconfiguration below:
+	//      ...
+	//	 	inputs:
+	//  		certificate-to-cross-sign-path: int-e6.cert.pem
+	//		certificate-profile:
+	//  		common-name: (FAKE) E5
+	//  		organization: (FAKE) Let's Encrypt
+	//      ...
+	//
+	if !bytes.Equal(toBeCrossSigned.RawSubject, lintCert.RawSubject) {
+		return fmt.Errorf("mismatch between toBeCrossSigned and lintCert RawSubject DER bytes: \"%x\" != \"%x\"", toBeCrossSigned.RawSubject, lintCert.RawSubject)
+	}
+	// CPS 7.1 Certificate Profile: Subordinate CA Certificate Validity Period
+	// Check that the cert has a valid validity period of "Up to 8 years". The
+	// validity period is computed inclusive of the whole final second indicated
+	// by notAfter.
+	parsedValidityDuration := lintCert.NotAfter.Add(time.Second).Sub(lintCert.NotBefore)
+	if time.Duration.Seconds(252288000) > parsedValidityDuration.Seconds() {
+		return fmt.Errorf("subordinate CA validity period is greater than the CPS allowed 8 calendar years (252288000 seconds): %.0f seconds\n", parsedValidityDuration.Seconds())
+	}
+	// BR 7.1.2.2.1 Cross-Certified Subordinate CA Validity
+	// The earlier of one day prior to the time of signing or the earliest
+	// notBefore date of the existing CA Certificate(s).
+	if lintCert.NotBefore.Before(issuer.NotBefore) {
+		return fmt.Errorf("subordinate CA NotBefore predates the issuer NotBefore")
+	}
+	if !bytes.Equal(issuer.RawSubject, lintCert.RawIssuer) {
+		return fmt.Errorf("mismatch between issuer RawSubject and lintCert RawIssuer DER bytes: \"%x\" != \"%x\"", issuer.RawSubject, lintCert.RawIssuer)
+	}
+	// BR 7.1.2.2.3 Cross-Certified Subordinate CA Extensions
+	if !reflect.DeepEqual(lintCert.ExtKeyUsage, toBeCrossSigned.ExtKeyUsage) {
```

Comment:
I'd love to do this without `reflect`, but this is probably the best we can do until we switch to go1.21 and can use `slices.Equal()`. I've added a note to https://github.com/letsencrypt/boulder/issues/7032 to follow up on this.